File: Assembly-FSharp/Assembly-FSharp/Entities/Puck.fs
namespace Styx.Entities

module PuckModule =
    open RageCure.StateUtils
    open Styx.Models

    type Message =
    | UpdateToLoose
    | UpdateToCarried

    type Puck =
        { rigidbodyModel    : RigidbodyModel
        ; transformModel    : TransformModel
        }

    type States =
        { puckLoose     : IState<Puck>
        ; puckCarried   : IState<Puck>
        }

    type PuckState = 
        { puck: Puck
        ; states : States
        ; stateMachine : StateMachine<Puck>
        }

    let focusPuck (puckState: PuckState) =
        puckState.puck

    let reduce (msg: Message) (puckState: PuckState) =
        match msg with
        | UpdateToLoose     -> puckState.stateMachine.ChangeState puckState.states.puckLoose    |> ignore; puckState 
        | UpdateToCarried   -> puckState.stateMachine.ChangeState puckState.states.puckCarried  |> ignore; puckState

    

File: Assets/_Scripts/Entities/Puck.cs
using RageCure.StateUtils;
using Styx.Models;
using UnityEngine;

namespace Styx.Entities.PuckModule
{
    public enum Message
    {
        UpdateToLoose,
        UpdateToCarried
    }

    public class Puck
    {
        public RigidbodyModel RigidbodyModel { get; set; }
        public TransformModel TransformModel { get; set; }

        public Rigidbody Rigidbody { get; set; }

        public void UpdateRigidbody()
        {
            Rigidbody.mass = RigidbodyModel.Mass;
            Rigidbody.drag = RigidbodyModel.Drag;
            Rigidbody.angularDrag = RigidbodyModel.AngularDrag;
        }

        public void SetPosition (Vector3 pos)
        {
            TransformModel.Transform.position = pos;
        }

        public void SetRotation (Vector3 rot)
        {
            TransformModel.Transform.rotation = Quaternion.Euler(rot);
        }
    }

    public class States
    {
        public State<PuckState> PuckLoose { get; set; }
        public State<PuckState> PuckCarried { get; set; }
    }

    public class PuckState
    {
        public Puck Puck { get; set; }
        public States States { get; set; }
        public StateMachine<PuckState> StateMachine { get; set; }

        public void Reduce (Message msg)
        {
            switch(msg)
            {
                case Message.UpdateToLoose:
                    StateMachine.ChangeState(this, States.PuckLoose);
                    break;
                case Message.UpdateToCarried:
                    StateMachine.ChangeState(this, States.PuckCarried);
                    break;
            }
        }
    }


}
